From 2c89e1da6163182b7e2f76f132f9221dae7eddf7 Mon Sep 17 00:00:00 2001
From: John Chung <john.chung@mic.com.tw>
Date: Wed, 11 Nov 2020 16:37:54 +0800
Subject: [PATCH] Fix SEL event issue for the power and reset button

Add the power button SEL event to long press
Fix the issue for the reset button would not log SEL event when host power off
Refactor the generate butoon SEL fucntion
---
 inc/button_handler.hpp |  7 ++++++
 src/button_handler.cpp | 63 ++++++++++++++++++++++----------------------------
 2 files changed, 34 insertions(+), 36 deletions(-)
 mode change 100644 => 100755 inc/button_handler.hpp

diff --git a/inc/button_handler.hpp b/inc/button_handler.hpp
old mode 100644
new mode 100755
index 7340c2f..1f3c007
--- a/inc/button_handler.hpp
+++ b/inc/button_handler.hpp
@@ -83,6 +83,13 @@ class Handler
     bool poweredOn() const;
 
     /**
+     * @brief Log the SEL for button event
+     *
+     * @param[in] eventData0 - event data 0 for SEL
+     */
+    void logButtonEvent(uint8_t eventData0) const;
+
+    /**
      * @brief Returns the service name for an object
      *
      * @param[in] path - the object path
diff --git a/src/button_handler.cpp b/src/button_handler.cpp
index 1c819b6..ff210ec 100755
--- a/src/button_handler.cpp
+++ b/src/button_handler.cpp
@@ -137,17 +137,36 @@ bool Handler::poweredOn() const
                sdbusplus::message::variant_ns::get<std::string>(state));
 }
 
-void Handler::powerPressed(sdbusplus::message::message& msg)
+void Handler::logButtonEvent(uint8_t eventData0) const
 {
-    auto transition = Host::Transition::On;
-
     uint16_t genid = 0x20;
     bool assert = 1;
     std::vector<uint8_t> eventData(selEvtDataMaxSize, 0xFF);
-    eventData[0] = 0;
+    eventData[0] = eventData0;
 
+    sdbusplus::message::message writeSEL = bus.new_method_call(
+        ipmiSELService, ipmiSELPath, ipmiSELAddInterface, "IpmiSelAdd");
+    writeSEL.append(ipmiSELAddMessage, sensorPath, eventData, assert,
+                    genid);
     try
     {
+        bus.call(writeSEL);
+    }
+    catch (SdBusError& e)
+    {
+        log<level::ERR>("Failed to log the button event",
+                        entry("ERROR=%s", e.what()));
+    }
+}
+
+void Handler::powerPressed(sdbusplus::message::message& msg)
+{
+    auto transition = Host::Transition::On;
+
+    try
+    {
+        logButtonEvent(0x00);
+
         if (poweredOn())
         {
             transition = Host::Transition::Off;
@@ -155,19 +174,6 @@ void Handler::powerPressed(sdbusplus::message::message& msg)
 
         log<level::INFO>("Handling power button press");
 
-        sdbusplus::message::message writeSEL = bus.new_method_call(
-            ipmiSELService, ipmiSELPath, ipmiSELAddInterface, "IpmiSelAdd");
-        writeSEL.append(ipmiSELAddMessage, sensorPath, eventData, assert,
-                    genid);
-        try
-        {
-            bus.call(writeSEL);
-        }
-        catch (sdbusplus::exception_t& e)
-        {
-            log<level::ERR>("call IpmiSelAdd failed");
-        }
-
     }
     catch (SdBusError& e)
     {
@@ -186,6 +192,8 @@ void Handler::longPowerPressed(sdbusplus::message::message& msg)
 
     try
     {
+        logButtonEvent(0x00);
+
         if (!poweredOn())
         {
             log<level::INFO>(
@@ -218,14 +226,10 @@ void Handler::longPowerPressed(sdbusplus::message::message& msg)
 
 void Handler::resetPressed(sdbusplus::message::message& msg)
 {
-
-    uint16_t genid = 0x20;
-    bool assert = 1;
-    std::vector<uint8_t> eventData(selEvtDataMaxSize, 0xFF);
-    eventData[0] = 0x02;
-
     try
     {
+        logButtonEvent(0x02);
+
         if (!poweredOn())
         {
             log<level::INFO>("Power is off so ignoring reset button press");
@@ -233,19 +237,6 @@ void Handler::resetPressed(sdbusplus::message::message& msg)
         }
 
         log<level::INFO>("Handling reset button press");
-
-        sdbusplus::message::message writeSEL = bus.new_method_call(
-            ipmiSELService, ipmiSELPath, ipmiSELAddInterface, "IpmiSelAdd");
-        writeSEL.append(ipmiSELAddMessage, sensorPath, eventData, assert,
-                    genid);
-        try
-        {
-            bus.call(writeSEL);
-        }
-        catch (sdbusplus::exception_t& e)
-        {
-            log<level::ERR>("call IpmiSelAdd failed");
-        }
     }
     catch (SdBusError& e)
     {
-- 
2.7.4

